src/lib/pubkey/dsa/dsa.cpp

```diff
@@ -23,10 +23,21 @@ namespace Botan {
 /*
 * DSA_PublicKey Constructor
 */
+
+DSA_PublicKey::DSA_PublicKey(const AlgorithmIdentifier& alg_id,
+                             const std::vector<uint8_t>& key_bits) :
+   DL_Scheme_PublicKey(alg_id, key_bits, DL_Group_Format::ANSI_X9_57)
+   {
+   BOTAN_ARG_CHECK(group_q().bytes() > 0, "Q parameter must be set for DSA");
+   }
+
+
 DSA_PublicKey::DSA_PublicKey(const DL_Group& grp, const BigInt& y1)
    {
    m_group = grp;
    m_y = y1;
+
+   BOTAN_ARG_CHECK(grp.q_bytes() > 0, "Q parameter must be set for DSA");
    }
```

Comment:
That is suppose to fix asserts in `PK_Verifier` for a DSA public key in the `cert` corpus/fuzzer. I left the assert untouched and instead introduce sanity checks at construction time of the DSA public key.
@randombit I could use a little help with this, as I'm not sure whether other things should also be validated here.

By the way: The currently still failing assert in the `pkcs8` fuzzer is related to a DSA private key:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007ff81db8e00e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007ff81dbc41ff libsystem_pthread.dylib`pthread_kill + 263
    frame #2: 0x00007ff81db0fd24 libsystem_c.dylib`abort + 123
    frame #3: 0x00000001009a9778 pkcs8`Botan::assertion_failure(char const*, char const*, char const*, char const*, int) + 696
    frame #4: 0x000000010048ed39 pkcs8`Botan::Montgomery_Exponentation_State::exponentiation(Botan::BigInt const&, unsigned long) const + 1001
    frame #5: 0x000000010048fc46 pkcs8`Botan::monty_execute(Botan::Montgomery_Exponentation_State const&, Botan::BigInt const&, unsigned long) + 22
    frame #6: 0x0000000100651218 pkcs8`Botan::DL_Group_Data::power_g_p(Botan::BigInt const&, unsigned long) const + 72
    frame #7: 0x000000010065127b pkcs8`Botan::DL_Group::power_g_p(Botan::BigInt const&, unsigned long) const + 59
    frame #8: 0x0000000100660e4f pkcs8`Botan::DSA_PrivateKey::DSA_PrivateKey(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 767
    frame #9: 0x000000010080913c pkcs8`std::__1::__unique_if<Botan::DSA_PrivateKey>::__unique_single std::__1::make_unique<Botan::DSA_PrivateKey, Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&>(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 76
    frame #10: 0x000000010080851b pkcs8`Botan::load_private_key(Botan::AlgorithmIdentifier const&, std::__1::vector<unsigned char, Botan::secure_allocator<unsigned char> > const&) + 1003
    frame #11: 0x000000010083fe17 pkcs8`Botan::PKCS8::(anonymous namespace)::load_key(Botan::DataSource&, std::__1::function<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()> const&, bool) + 471
    frame #12: 0x0000000100840587 pkcs8`Botan::PKCS8::load_key(Botan::DataSource&) + 231
    frame #13: 0x0000000100002c8f pkcs8`fuzz(unsigned char const*, unsigned long) + 255
    frame #14: 0x0000000100002b8a pkcs8`LLVMFuzzerTestOneInput + 26
    frame #15: 0x0000000100003483 pkcs8`(anonymous namespace)::fuzz_files(char**) + 899
    frame #16: 0x0000000100002eb7 pkcs8`main + 231
    frame #17: 0x0000000101df551e dyld`start + 462
```

Reply:
I think this assert would trigger if you load a DSA key with a secret key that is greater than the group order.

This likely affects other key types as well.

I think we need to rethink key validation somewhat. Right now all validation is dumped on `check_key`, because some of the checks (testing primality and ECC point validity) are expensive. But most are simple range checks and could be cheaply done during construction or deserialization.

This is resolved with 1735f95 right?